pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,277 @@
+		}
+	}
+
+	if len(address) == 0 {
+		return nil, util2.NoFilteredPod{
+			Desc: "no pod has the corresponding named port, perhaps pod has not been created yet",
+		}
+	}
+	return util.MergeAddressByPort(address), nil
+}
+
+// Check port name and protocol, only when the pod is really running, and it does have effective ip.
+func (service *SecurityPolicyService) resolvePodPort(pod *v1.Pod, spPort *v1alpha1.SecurityPolicyPort) (
+	[]util.Address, error) {
+	var addr []util.Address
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			log.V(2).Info("resolvePodPort", "namespace", pod.Namespace, "pod_name", pod.Name,
+				"port_name", port.Name, "containerPort", port.ContainerPort,
+				"protocol", port.Protocol, "podIP", pod.Status.PodIP)
+			if port.Name == spPort.Port.String() && port.Protocol == spPort.Protocol {
+				if pod.Status.Phase != "Running" {
+					errMsg := fmt.Sprintf("pod %s/%s is not running", pod.Namespace, pod.Name)
+					return nil, util2.PodNotRunning{Desc: errMsg}
+				}
+				if pod.Status.PodIP == "" {
+					errMsg := fmt.Sprintf("pod %s/%s ip not initialized", pod.Namespace, pod.Name)
+					return nil, util2.PodIPNotFound{Desc: errMsg}
+				}
+				addr = append(addr, util.Address{Port: int(port.ContainerPort), IPs: []string{pod.Status.PodIP}})
+			}
+		}
+	}
+	return addr, nil
+}
+
+// Build an ip set group for NSX.
+func (service *SecurityPolicyService) buildRuleIPGroup(obj *model.Rule, ips []string) *model.Group {
+	ipGroup := model.Group{}
+
+	policyGroupID := fmt.Sprintf("%s_ipset", *obj.Id)
+	ipGroup.Id = &policyGroupID
+	policyGroupName := fmt.Sprintf("%s-ipset", *obj.DisplayName)
+	ipGroup.DisplayName = &policyGroupName
+
+	addresses := data.NewListValue()
+	for _, ip := range ips {
+		addresses.Add(data.NewStringValue(ip))
+	}
+
+	blockExpression := data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"resource_type": data.NewStringValue("IPAddressExpression"),
+			"ip_addresses":  addresses,
+		},
+	)
+	ipGroup.Expression = append(ipGroup.Expression, blockExpression)
+	return &ipGroup
+}
+
+// Different direction rule decides different target of the traffic, we should carefully get
+// the destination pod selector and namespaces.
+func (service *SecurityPolicyService) getPodSelector(obj *v1alpha1.SecurityPolicy,
+	rule *v1alpha1.SecurityPolicyRule) (*client.ListOptions, []string, error) {
+
+	// Port means the target of traffic, so we should select the pod by rule direction,
+	// as for IN direction, we judge by rule's or policy's AppliedTo,
+	// as for OUT direction, then by rule's destinations.
+	// LabelSelect may return multiple namespaces
+	var namespaces []string
+	selector := client.ListOptions{}
+	var labelSelector labels.Selector
+	ruleDirection, err := getRuleDirection(rule)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if ruleDirection == "IN" {
+		if len(rule.AppliedTo) > 0 {
+			for _, target := range rule.AppliedTo {
```

Comment:
As we know policy's AppliedTo will take precedence over rule's AppliedTo? Should we check  policy's
AppliedTo firstly? When policy's AppliedTo will override rule's AppliedTo?

Reply:
Yes, I think it is more reasonable, @heypnus how do you think.